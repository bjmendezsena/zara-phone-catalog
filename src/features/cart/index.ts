export * from "./context";
export * from "./helpers";
export * from "./types";
export * from "./components";
