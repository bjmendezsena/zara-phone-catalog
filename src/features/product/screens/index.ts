export * from "./ProductDetails";
export * from "./ProductGrid";

