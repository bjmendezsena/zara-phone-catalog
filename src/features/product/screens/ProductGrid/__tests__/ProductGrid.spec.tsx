import {
  getHtml,
  mockProductListResponse,
  renderWithProviders,
  waitFor,
  userEvent,
} from "@/tests";
import { ProductGrid } from "@/features/product";

describe(`<${ProductGrid.name}/>`, () => {
  const getComponent = () => renderWithProviders(<ProductGrid />);

  it("renders without crashing", () => {
    const { container } = getComponent();
    expect(getHtml(container)).toMatchSnapshot();
  });

  it("displays the correct number of products", async () => {
    const { getAllByTestId } = getComponent();

    await waitFor(() => {
      const productItems = getAllByTestId("product-card");
      expect(productItems).toHaveLength(mockProductListResponse.length);
      expect(`${mockProductListResponse.length} Results`).toBe(
        getAllByTestId("search-results-count")[0].textContent
      );
    });
  });

  it("shows loading state initially", () => {
    const { getByText } = getComponent();
    expect(getByText("Loading...")).toBeInTheDocument();
  });

  it("Can search for products", async () => {
    const { getByTestId, getByText } = getComponent();
    const searchInput = getByTestId("search-input");
    const search = mockProductListResponse[0].name;
    const filtreredProducts = mockProductListResponse.filter((product) =>
      product.name.toLowerCase().includes(search.toLowerCase())
    );
    await userEvent.type(searchInput, search);
    expect(searchInput).toHaveValue(search);

    await waitFor(() => {
      expect(getByTestId("clear-search-button")).toBeInTheDocument();
      expect(getByText(search)).toBeInTheDocument();

      expect(`${filtreredProducts.length} Results`).toBe(
        getByTestId("search-results-count").textContent
      );
    });
  });

  it("Can clear input for products", async () => {
    const { getByTestId, getByText } = getComponent();
    const searchInput = getByTestId("search-input");
    const search = mockProductListResponse[0].name;
    const filtreredProducts = mockProductListResponse.filter((product) =>
      product.name.toLowerCase().includes(search.toLowerCase())
    );
    await userEvent.type(searchInput, search);
    expect(searchInput).toHaveValue(search);

    await waitFor(() => {
      expect(getByTestId("clear-search-button")).toBeInTheDocument();
      expect(getByText(search)).toBeInTheDocument();

      expect(`${filtreredProducts.length} Results`).toBe(
        getByTestId("search-results-count").textContent
      );
    });

    await userEvent.click(getByTestId("clear-search-button"));
    expect(searchInput).toHaveValue("");
    await waitFor(() => {
      expect(getByTestId("search-results-count").textContent).toBe(
        `${mockProductListResponse.length} Results`
      );
    });
  });
});
