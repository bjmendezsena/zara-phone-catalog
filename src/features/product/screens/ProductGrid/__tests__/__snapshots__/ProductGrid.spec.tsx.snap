// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<ProductGrid/> renders without crashing 1`] = `
"<div>
  <div
    class="w-full h-full flex flex-col items-center justify-center p-4 sm:p-8 gap-8"
  >
    <div
      aria-label="Formulario de bÃºsqueda de productos"
      class="flex flex-col w-full gap-4 group"
    >
      <div
        class="relative w-full"
      >
        <input
          autocomplete="off"
          class="w-full p-2 border-b border-primary focus:outline-none placeholder:text-placeholder peer pr-8"
          data-testid="search-input"
          name="search"
          placeholder="Buscar productos..."
          type="text"
        />
      </div>
      <span
        class="col-span-full text-secondary mb-4 peer-focus:opacity-0 opacity-0"
        data-testid="search-results-count"
      >
        0
         Results
      </span>
    </div>
    <div
      aria-label="Lista de productos"
      class="grid grid-cols-1 xs:grid-cols-2 sm:grid-cols-3 md:grid-cols-4 lg:grid-cols-5 w-full"
    >
      <div
        class="col-span-full flex items-center justify-center"
      >
        <span
          class="col-span-full text-center text-secondary font-light mb-4"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>"
`;
